src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Date.cs

```diff
@@ -42,103 +42,73 @@ public static void WriteDateTimeOffsetTrimmed(Span<byte> buffer, DateTimeOffset
         //   2017-06-12T05:30:45                  (interpreted as local time wrt to current time zone)
         public static void TrimDateTimeOffset(Span<byte> buffer, out int bytesWritten)
         {
+            const int maxLenNoOffset = JsonConstants.MaximumFormatDateTimeLength;
+            const int maxLenWithZ = maxLenNoOffset + 1;
+            const int maxLenWithOffset = JsonConstants.MaximumFormatDateTimeOffsetLength;
+
             // Assert buffer is the right length for:
             // YYYY-MM-DDThh:mm:ss.fffffff (JsonConstants.MaximumFormatDateTimeLength)
             // YYYY-MM-DDThh:mm:ss.fffffffZ (JsonConstants.MaximumFormatDateTimeLength + 1)
             // YYYY-MM-DDThh:mm:ss.fffffff(+|-)hh:mm (JsonConstants.MaximumFormatDateTimeOffsetLength)
-            Debug.Assert(buffer.Length == JsonConstants.MaximumFormatDateTimeLength ||
-                buffer.Length == (JsonConstants.MaximumFormatDateTimeLength + 1) ||
-                buffer.Length == JsonConstants.MaximumFormatDateTimeOffsetLength);
-
-            uint digit7 = buffer[26] - (uint)'0';
-            uint digit6 = buffer[25] - (uint)'0';
-            uint digit5 = buffer[24] - (uint)'0';
-            uint digit4 = buffer[23] - (uint)'0';
-            uint digit3 = buffer[22] - (uint)'0';
-            uint digit2 = buffer[21] - (uint)'0';
-            uint digit1 = buffer[20] - (uint)'0';
-            uint fraction = (digit1 * 1_000_000) + (digit2 * 100_000) + (digit3 * 10_000) + (digit4 * 1_000) + (digit5 * 100) + (digit6 * 10) + digit7;
-
-            // The period's index
-            int curIndex = 19;
-
-            if (fraction > 0)
+            Debug.Assert(buffer.Length == maxLenNoOffset ||
+                buffer.Length == maxLenWithZ ||
+                buffer.Length == maxLenWithOffset);
+
+            // Find the position after the last significant digit in seconds fraction.
+            int curIndex;
+            if (buffer[maxLenNoOffset - 1] == '0')
```

Comment:
There's no question that the existing implementation is doing unnecessary work, but that being said I do find the formatting of this particular section to be slightly disagreeable. This is not a hot path method so I'm not sure it's worth unrolling. I'd be interested to see performance numbers with your `curIndex` calculation logic refactored into a loop.

Reply:
The loop version is only slightly slower. I would say this path can get pretty hot depending on the payload. But I agree - will change.

Reply:
Thanks